python/test_infra/ttir_builder.py

```diff
@@ -0,0 +1,342 @@
+    seed: int = None
+
+    @staticmethod
+    def generate_random(shape: Shape) -> Golden:
+        """Generates random golden tensor with `shape`."""
+        seed = TTIRBuilder.generate_random_seed()
+        tensor = TTIRBuilder.generate_random_tensor(shape, seed)
+        return Golden(tensor, seed)
+
+    def __repr__(self) -> str:
+        s = f"\nRandom seed: {self.seed}" if self.seed else ""
+        s += f"\nGolden tensor:\n{self.tensor}"
+        return s
+
+
+class TTIRBuilder:
+    """Builder class providing API for creating TTIR ops."""
+
+    def __init__(self, ctx: Context, location: Location):
+        self._ctx = ctx
+        self._loc = location
+
+        tt.register_dialect(self._ctx)
+        ttir.register_dialect(self._ctx)
+
+        # Dictionary to store Golden for each Operand we encounter in MLIR graph.
+        self._goldens: Dict[Operand, Golden] = {}
+
+    # ----- Public helpers -----
+
+    @staticmethod
+    def generate_random_seed() -> int:
+        """Generates random integer in range [0, 10)."""
+        return torch.randint(low=0, high=10, size=())
+
+    @staticmethod
+    def generate_random_tensor(shape: Shape, seed: int) -> torch.Tensor:
+        """Generates random tensor of shape `shape`, using `seed` to seed torch random generator."""
+        return torch.randn(shape, generator=torch.manual_seed(seed))
+
+    def print_goldens(self) -> None:
+        """
+        Prints saved operands and their respective goldens in descriptive form which follows SSA
+        ordering from MLIR graph.
+        """
+        i = 0
+        for operand, golden in self._goldens.items():
+            operand_name = self._get_name(operand)
+            if "arg" in operand_name:
+                print(f"Func arg: {operand_name}", golden)
+            else:
+                print(f"%{i}: {operand_name}", golden)
+                i += 1
+
+    def get_shape(self, input: Operand) -> Shape:
+        """Retrieves shape of operand which is expected to be a shaped type."""
+        return self._get_type(input).shape
+
+    # ----- Private helpers -----
+
+    @staticmethod
+    def _get_name(operand: Operand) -> str:
+        """Retrieves descriptive operand name."""
+        # Try to call get_name() if it exists, otherwise return operand.name.
+        return getattr(operand, "get_name", lambda: None)() or getattr(
+            operand, "name", None
+        )
+
+    def _store_golden(self, operand: Operand, golden: Golden) -> None:
+        """Maps `operand` to `golden`."""
+        assert (
+            self._goldens.get(operand) == None
+        ), f"Golden for {operand} already exists."
+        self._goldens[operand] = golden
+
+    def _get_golden(self, operand: Operand) -> Golden:
+        """
+        Retrieves stored golden for `operand`.
+
+        If no golden is found for `operand`, generates a random one.
+        """
+        golden = self._goldens.get(operand)
+
+        if golden is None:
```

Comment:
Shouldn't this be an assert?  I feel like it's a bug if `operand` doesn't have a golden.

Reply:
Good feeling, but I did it on purpose to hide a bit of complexity from op factories in special case where operand is not a result of another op, but rather a surrounding MLIR (and python) function arg.

For example

```python
class TTIRBuilder:
    ...
    def multiply(self, in0: Operand, in1: Operand) -> OpView:
        """Convenience wrapper constructing `ttir.MultiplyOp`."""
        assert self.get_shape(in0) == self.get_shape(
            in1
        ), "Elementwise `ttir.multiply` op expects inputs of same shape."

        with self._ctx, self._loc:
            output = self.empty(self.get_shape(in0))

            op = ttir.MultiplyOp(
                [self._get_type(output)],
                [in0, in1],
                [output],
                self._get_operand_constraint_attr(3),
            )

            golden = torch.multiply(
                self._get_golden_tensor(in0), self._get_golden_tensor(in1)
            )
            self._store_golden(op, Golden(golden))

            return op


@compile_as_mlir_module((64, 64), (64, 64))
def test_multiply(in0: Operand, in1: Operand, builder: TTIRBuilder):
    return builder.multiply(in0, in1)
```

with corresponding

```mlir
#any = #tt.operand_constraint<system|dram|l1|scalar|tile|none|interleaved|single_bank|height_sharded|width_sharded|block_sharded|any_layout|any|any_device|any_device_tile|l1_block_sharded>
module {
  func.func @test_multiply(%arg0: tensor<64x64xf32>, %arg1: tensor<64x64xf32>) -> tensor<64x64xf32> {
    %0 = tensor.empty() : tensor<64x64xf32>
    %1 = "ttir.multiply"(%arg0, %arg1, %0) <{operandSegmentSizes = array<i32: 2, 1>, operand_constraints = [#any, #any, #any]}> : (tensor<64x64xf32>, tensor<64x64xf32>, tensor<64x64xf32>) -> tensor<64x64xf32>
    return %1 : tensor<64x64xf32>
  }
}
```

`in0` and `in1` are func args, they are not instantiated beforehand. In order to have a clean

```python
golden = torch.multiply(
    self._get_golden_tensor(in0), self._get_golden_tensor(in1)
)
```

in `TTIR.multiply` method, the complexities of `self._goldens.get(in0) == None` checking and creating a golden, which we would need to have in every TTIR op factory method, are hidden away in `self._get_golden` under this condition. 

It does have a bit of unexpected, though explained with a comment, side effect, I agree, but I don't see a cleaner way of doing it. Please suggest if you see a nicer way.

I added an assert which makes sure only operands that are wrapping MLIR's function args can reach that part of code.

Reply:
Shouldn't they be generated/registered as part of the wrapper?

```python
def decorated_func(*inputs):
    for i in inputs:
        builder._store_golden(i, generate_golden...)
    return test_fn(*inputs, builder=builder)
```

It seems a lot safer to make this an assert, otherwise we can just silently be getting new goldens for values that came from somewhere and it's hard to debug.  The assert enforces that the value + golden was created and explicitly tracked.

Reply:
You are absolutely right. Fixed.